File: go/column_prefetch.go
package nxs

import "fmt"

// PrefetchColumn issues a single range fetch for a columnar column buffer (§7.4).
// Row page cache is not used. Idempotent per column slot.
func (r *Reader) PrefetchColumn(key string) error {
	if r.layout != LayoutColumnar {
		return fmt.Errorf("ERR_LAYOUT: prefetch_column requires columnar layout")
	}
	slot, ok := r.keyIndex[key]
	if !ok {
		return fmt.Errorf("ERR_KEY_NOT_FOUND: %q", key)
	}
	r.colMu.Lock()
	defer r.colMu.Unlock()
	if r.colWarmed[slot] {
		return nil
	}
	off := int64(r.colBufOff[slot])
	ln := int64(r.colBufLen[slot])
	blob, err := r.colFetchRange(off, ln)
	if err != nil {
		return err
	}
	if int(off)+len(blob) > len(r.data) {
		r.colOverlay[slot] = blob
	}
	r.colWarmed[slot] = true
	r.colFetches++
	return nil
}

func (r *Reader) initColumnPrefetch(cfg readerConfig) {
	data := r.data
	fetchRange := cfg.fetchRange
	if fetchRange == nil {
		fetchRange = func(off, length int64) ([]byte, error) {
			return sliceInt64(data, off, length)
		}
	}
	r.colFetchRange = fetchRange
	r.colWarmed = make(map[int]bool)
	r.colOverlay = make(map[int][]byte)
}

func (r *Reader) columnSector(slot int) ([]byte, error) {
	if slot < 0 || slot >= len(r.colBufOff) {
		return nil, fmt.Errorf("ERR_KEY_NOT_FOUND")
	}
	off := int(r.colBufOff[slot])
	length := int(r.colBufLen[slot])
	r.colMu.Lock()
	overlay, warmed := r.colOverlay[slot], r.colWarmed[slot]
	r.colMu.Unlock()
	if warmed && len(overlay) > 0 {
		if len(overlay) < length {
			return nil, fmt.Errorf("ERR_OUT_OF_BOUNDS: column overlay")
		}
		return overlay[:length], nil
	}
	if off+length > len(r.data) {
		return nil, fmt.Errorf("ERR_OUT_OF_BOUNDS: column buffer")
	}
	return r.data[off : off+length], nil
}

File: csharp/ColumnPrefetch.cs
// Columnar column-buffer warmup (Adaptive-prefetch-spec §7.4).

using System;
using System.Collections.Generic;

namespace Nxs;

internal sealed class ColumnWarmState
{
    private readonly byte[] _data;
    private readonly Func<long, long, byte[]> _fetch;
    private readonly bool _customFetch;
    private readonly HashSet<int> _warmed = new();
    private readonly Dictionary<int, byte[]> _overlay = new();
    private readonly object _lock = new();
    public int Fetches { get; private set; }

    public ColumnWarmState(byte[] data, Func<long, long, byte[]>? fetchRange, bool customFetch = false)
    {
        _data = data;
        _customFetch = customFetch || fetchRange != null;
        _fetch = fetchRange ?? ((off, len) =>
        {
            long end = off + len;
            if (off < 0 || end > data.Length)
                throw new NxsException("ERR_OUT_OF_BOUNDS", $"column fetch [{off}, {end})");
            int lenInt = checked((int)len);
            var blob = new byte[lenInt];
            Array.Copy(data, (int)off, blob, 0, lenInt);
            return blob;
        });
    }

    public void PrefetchColumn(int slot, ulong[] colOff, ulong[] colLen)
    {
        long off;
        long len;
        lock (_lock)
        {
            if (_warmed.Contains(slot)) return;
            off = (long)colOff[slot];
            len = (long)colLen[slot];
            if (off < 0 || len < 0)
                throw new NxsException("ERR_OUT_OF_BOUNDS", "column buffer");
            if (!_customFetch && off + len > _data.Length)
                throw new NxsException("ERR_OUT_OF_BOUNDS", "column buffer");
        }

        byte[] blob = _fetch(off, len);

        lock (_lock)
        {
            if (_warmed.Contains(slot)) return;
            if (off + blob.Length > _data.Length)
                _overlay[slot] = blob;
            _warmed.Add(slot);
            Fetches++;
        }
    }

    public byte[] Sector(int slot, ulong[] colOff, ulong[] colLen)
    {
        int length = (int)colLen[slot];
        lock (_lock)
        {
            if (_overlay.TryGetValue(slot, out byte[]? o) && o.Length >= length)
            {
                var slice = new byte[length];
                Array.Copy(o, 0, slice, 0, length);
                return slice;
            }
        }
        int off = (int)colOff[slot];
        if (off >= 0 && off + length <= _data.Length)
        {
            var sector = new byte[length];
            Array.Copy(_data, off, sector, 0, length);
            return sector;
        }
        throw new NxsException("ERR_OUT_OF_BOUNDS", "column buffer");
    }
}
